File: runtime-rust/src/metrics.rs
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::time::Duration;

lazy_static::lazy_static! {
    pub static ref GLOBAL_METRICS: Arc<MetricsCollector> = Arc::new(MetricsCollector::new());
}

pub struct MetricsCollector {
    // key: (method, path, status) -> count
    requests: RwLock<HashMap<(String, String, u16), u64>>,
    // key: (method, path) -> count
    errors: RwLock<HashMap<(String, String), u64>>,
    // key: (method, path) -> list of durations in seconds
    durations: RwLock<HashMap<(String, String), Vec<f64>>>,
}

impl MetricsCollector {
    pub fn new() -> Self {
        MetricsCollector {
            requests: RwLock::new(HashMap::new()),
            errors: RwLock::new(HashMap::new()),
            durations: RwLock::new(HashMap::new()),
        }
    }

    pub fn record(&self, method: &str, path: &str, status: u16, duration: Duration) {
        let dur_secs = duration.as_secs_f64();
        
        let mut reqs = self.requests.write().unwrap();
        *reqs.entry((method.to_string(), path.to_string(), status)).or_insert(0) += 1;

        let mut durs = self.durations.write().unwrap();
        durs.entry((method.to_string(), path.to_string())).or_default().push(dur_secs);

        if status >= 400 {
            let mut errs = self.errors.write().unwrap();
            *errs.entry((method.to_string(), path.to_string())).or_insert(0) += 1;
        }
    }

    pub fn format_prometheus(&self) -> String {
        let mut lines = Vec::new();

        lines.push("# HELP helix_requests_total Total number of RPC requests.".to_string());
        lines.push("# TYPE helix_requests_total counter".to_string());
        {
            let reqs = self.requests.read().unwrap();
            for ((method, path, status), count) in reqs.iter() {
                lines.push(format!(
                    "helix_requests_total{{method=\"{}\",path=\"{}\",status=\"{}\"}} {}",
                    method, path, status, count
                ));
            }
        }

        lines.push("".to_string());
        lines.push("# HELP helix_errors_total Total number of RPC errors.".to_string());
        lines.push("# TYPE helix_errors_total counter".to_string());
        {
            let errs = self.errors.read().unwrap();
            for ((method, path), count) in errs.iter() {
                lines.push(format!(
                    "helix_errors_total{{method=\"{}\",path=\"{}\"}} {}",
                    method, path, count
                ));
            }
        }

        let buckets = vec![0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0];
        lines.push("".to_string());
        lines.push("# HELP helix_request_duration_seconds Request latency histogram.".to_string());
        lines.push("# TYPE helix_request_duration_seconds histogram".to_string());
        {
            let durs = self.durations.read().unwrap();
            for ((method, path), durations) in durs.iter() {
                let total: f64 = durations.iter().sum();
                let count = durations.len();
                
                for &le in &buckets {
                    let bucket_count = durations.iter().filter(|&&d| d <= le).count();
                    lines.push(format!(
                        "helix_request_duration_seconds_bucket{{method=\"{}\",path=\"{}\",le=\"{}\"}} {}",
                        method, path, le, bucket_count
                    ));
                }
                lines.push(format!(
                    "helix_request_duration_seconds_bucket{{method=\"{}\",path=\"{}\",le=\"+Inf\"}} {}",
                    method, path, count
                ));
                lines.push(format!(
                    "helix_request_duration_seconds_sum{{method=\"{}\",path=\"{}\"}} {:.6}",
                    method, path, total
                ));
                lines.push(format!(
                    "helix_request_duration_seconds_count{{method=\"{}\",path=\"{}\"}} {}",
                    method, path, count
                ));
            }
        }

        lines.join("\n") + "\n"
    }
}

impl Default for MetricsCollector {
    fn default() -> Self {
        Self::new()
    }
}

File: runtimes/python/helix_rt/metrics.py
"""
helix_rt.metrics
~~~~~~~~~~~~~~~~
Prometheus-compatible metrics endpoint for the Python Helix runtime.

Exposes a ``/metrics`` scrape endpoint in the Prometheus text exposition format.
Tracks request counts, error counts, and latency histograms per route.

Usage::

    from helix_rt.metrics import MetricsCollector

    collector = MetricsCollector()
    server.add_middleware(collector.middleware())
    server.register_route("GET", "/metrics", collector.handler)
    server.start()

Optional: install ``prometheus_client`` for richer output:
    pip install prometheus_client
"""

from __future__ import annotations

import time
from collections import defaultdict
from typing import Callable

from aiohttp import web


class MetricsCollector:
    """
    Lightweight, zero-dependency Prometheus metrics collector.

    Tracks per-route:
    - helix_requests_total{method, path, status}
    - helix_request_duration_seconds{method, path, le} (histogram)
    - helix_errors_total{method, path}
    """

    # Histogram bucket boundaries in seconds (matches Prometheus defaults)
    BUCKETS = [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0]

    def __init__(self) -> None:
        self._requests:  dict[tuple, int]         = defaultdict(int)
        self._errors:    dict[tuple, int]          = defaultdict(int)
        self._durations: dict[tuple, list[float]]  = defaultdict(list)

    def _record(self, method: str, path: str, status: int, duration: float) -> None:
        key = (method.upper(), path, status)
        self._requests[key] += 1
        self._durations[(method.upper(), path)] += [duration]
        if status >= 500:
            self._errors[(method.upper(), path)] += 1

    def middleware(self) -> Callable:
        """Return an aiohttp middleware that records per-request metrics."""
        collector = self

        @web.middleware
        async def _middleware(request: web.Request, handler: Callable) -> web.StreamResponse:
            start   = time.monotonic()
            status  = 500
            try:
                response = await handler(request)
                status  = response.status
                return response
            finally:
                duration = time.monotonic() - start
                collector._record(request.method, request.path, status, duration)

        return _middleware

    async def handler(self, _body: dict) -> web.Response:
        """Handler registered as GET /metrics — returns Prometheus text format."""
        lines = [
            "# HELP helix_requests_total Total number of RPC requests.",
            "# TYPE helix_requests_total counter",
        ]
        for (method, path, status), count in self._requests.items():
            lines.append(
                f'helix_requests_total{{method="{method}",path="{path}",status="{status}"}} {count}'
            )

        lines += [
            "",
            "# HELP helix_errors_total Total number of 5xx errors.",
            "# TYPE helix_errors_total counter",
        ]
        for (method, path), count in self._errors.items():
            lines.append(
                f'helix_errors_total{{method="{method}",path="{path}"}} {count}'
            )

        lines += [
            "",
            "# HELP helix_request_duration_seconds Request latency histogram.",
            "# TYPE helix_request_duration_seconds histogram",
        ]
        for (method, path), durations in self._durations.items():
            total   = sum(durations)
            count   = len(durations)
            for le in self.BUCKETS:
                bucket_count = sum(1 for d in durations if d <= le)
                lines.append(
                    f'helix_request_duration_seconds_bucket{{method="{method}",path="{path}",le="{le}"}} {bucket_count}'
                )
            lines.append(
                f'helix_request_duration_seconds_bucket{{method="{method}",path="{path}",le="+Inf"}} {count}'
            )
            lines.append(
                f'helix_request_duration_seconds_sum{{method="{method}",path="{path}"}} {total:.6f}'
            )
            lines.append(
                f'helix_request_duration_seconds_count{{method="{method}",path="{path}"}} {count}'
            )

        body = "\n".join(lines) + "\n"
        return web.Response(
            status=200,
            content_type="text/plain; version=0.0.4",
            body=body.encode(),
        )
